# pandas 와 해보기

벡터 = 리스트(list)

## 벡터의 덧셈을 하는 함수 만들기

In [1]:
def v_add(u,v):    
    '''
    벡터의 덧셈
    입력값: 더하고자 하는 벡터 u,v
    출력값: 벡터 u, v 의 더한 결과 w
    '''
    n = len(u)
    w = []
    
    for i in range(0, n):
        val = u[i] + v[i]
        w.append(val)
        
    return w

In [2]:
u = [1, 2, 3]
v = [4, 5, 6]
v_add(u,v)

[5, 7, 9]

## 파일로 실습해보기

파일: mall_customer.csv

In [3]:
import pandas as pd

In [4]:
df = pd.read_csv('./data/mall_customer.csv')

In [5]:
df

CustomerID  Gender  Age  Annual Income  Spending Score
0             1       1   19             15              39
1             2       1   21             15              81
2             3       2   20             16               6
3             4       2   23             16              77
4             5       2   31             17              40
..          ...     ...  ...            ...             ...
195         196       2   35            120              79
196         197       2   45            126              28
197         198       1   32            126              74
198         199       1   32            137              18
199         200       1   30            137              83

[200 rows x 5 columns]

In [6]:
df.describe()

CustomerID      Gender         Age  Annual Income  Spending Score
count  200.000000  200.000000  200.000000     200.000000      200.000000
mean   100.500000    1.560000   38.850000      60.560000       50.200000
std     57.879185    0.497633   13.969007      26.264721       25.823522
min      1.000000    1.000000   18.000000      15.000000        1.000000
25%     50.750000    1.000000   28.750000      41.500000       34.750000
50%    100.500000    2.000000   36.000000      61.500000       50.000000
75%    150.250000    2.000000   49.000000      78.000000       73.000000
max    200.000000    2.000000   70.000000     137.000000       99.000000

## 데이터 안의 실제 값에 접근해보기 (iloc[i,j])

스칼라에 접근하기

In [7]:
df.iloc[2,2]

# 실제로 파이썬에서는 3행 3열을 가져오게 된다

20

벡터를 불러오기

In [8]:
df["Age"]

0      19
1      21
2      20
3      23
4      31
       ..
195    35
196    45
197    32
198    32
199    30
Name: Age, Length: 200, dtype: int64

열을 더하기

In [9]:
df["Age"] + df["Annual Income"] # a + b 와 같다

0       34
1       36
2       36
3       39
4       48
      ... 
195    155
196    171
197    158
198    169
199    167
Length: 200, dtype: int64

## 전치행렬 구하기 (transpose)

In [10]:
import numpy as np

In [11]:
df_t = np.transpose(df)

In [12]:
df_t

0    1    2    3    4    5    6    7    8    9    ...  190  \
CustomerID        1    2    3    4    5    6    7    8    9   10  ...  191   
Gender            1    1    2    2    2    2    2    2    1    2  ...    2   
Age              19   21   20   23   31   22   35   23   64   30  ...   34   
Annual Income    15   15   16   16   17   17   18   18   19   19  ...  103   
Spending Score   39   81    6   77   40   76    6   94    3   72  ...   23   

                191  192  193  194  195  196  197  198  199  
CustomerID      192  193  194  195  196  197  198  199  200  
Gender            2    1    2    2    2    2    1    1    1  
Age              32   33   38   47   35   45   32   32   30  
Annual Income   103  113  113  120  120  126  126  137  137  
Spending Score   69    8   91   16   79   28   74   18   83  

[5 rows x 200 columns]

## X^T  * X : 하기

.dot() : 곱하기

In [13]:
df_t.dot(df)

CustomerID  Gender     Age  Annual Income  Spending Score
CustomerID         2686700   31027  776579        1512980         1013135
Gender               31027     536   12037          18748           15811
Age                 776579   12037  340696         469646          366564
Annual Income      1512980   18748  469646         870780          609359
Spending Score     1013135   15811  366564         609359          636712

짜잔! 대칭행렬이 나왔습니다.

많이 나오는 이유, 대칭행렬을 만들기 위함이고, 이게 corrvariance matrix의 근간이다.


In [14]:
# pd.set_option('display.max.colwidth', 50)
# df_t

# pandas 없이 해보기

In [15]:
import csv

In [16]:
raw = []

with open('./data/mall_customer.csv', 'r', encoding = 'utf-8') as f:
    rdr = csv.reader(f)
    for line in rdr:
        raw.append(line)

csv 함수에 reader라는 함수가 있다. with open이라는 것의 용법.

행이 리스트 형태로 들어가 있다. 개발을 할 때는 행 중심이라는 것이 그 말.

그리고 작은 따옴표로 들어간 것, 전부 object로 들어가 있는 상황이라 숫자로 변환해주어야 한다.

In [17]:
raw

[['CustomerID', 'Gender', 'Age', 'Annual Income', 'Spending Score'],
 ['1', '1', '19', '15', '39'],
 ['2', '1', '21', '15', '81'],
 ['3', '2', '20', '16', '6'],
 ['4', '2', '23', '16', '77'],
 ['5', '2', '31', '17', '40'],
 ['6', '2', '22', '17', '76'],
 ['7', '2', '35', '18', '6'],
 ['8', '2', '23', '18', '94'],
 ['9', '1', '64', '19', '3'],
 ['10', '2', '30', '19', '72'],
 ['11', '1', '67', '19', '14'],
 ['12', '2', '35', '19', '99'],
 ['13', '2', '58', '20', '15'],
 ['14', '2', '24', '20', '77'],
 ['15', '1', '37', '20', '13'],
 ['16', '1', '22', '20', '79'],
 ['17', '2', '35', '21', '35'],
 ['18', '1', '20', '21', '66'],
 ['19', '1', '52', '23', '29'],
 ['20', '2', '35', '23', '98'],
 ['21', '1', '35', '24', '35'],
 ['22', '1', '25', '24', '73'],
 ['23', '2', '46', '25', '5'],
 ['24', '1', '31', '25', '73'],
 ['25', '2', '54', '28', '14'],
 ['26', '1', '29', '28', '82'],
 ['27', '2', '45', '28', '32'],
 ['28', '1', '35', '28', '61'],
 ['29', '2', '40', '29', '31'],
 ['30', '2', '23

## 행렬 살피기

In [18]:
n_row = len(raw)
n_col = len(raw[0])

In [19]:
print(n_row)
print(n_col)

201
5


데이터를 불러왔으면 크기를 확인하는 습관을 들여야 한다.

201로 1이 더 나온 이유는 컬럼 행 때문에 더 나온 것.

## 기존 데이터를 실수형태로 바꾸기

In [20]:
data = []
for i in range(1, n_row):     # 0번 행은 제목 행이라 안쓴다.
    row = []
    for j in range(0, n_col):
        val = float(raw[i][j])
        row.append(val)
    data.append(row)

In [21]:
data

[[1.0, 1.0, 19.0, 15.0, 39.0],
 [2.0, 1.0, 21.0, 15.0, 81.0],
 [3.0, 2.0, 20.0, 16.0, 6.0],
 [4.0, 2.0, 23.0, 16.0, 77.0],
 [5.0, 2.0, 31.0, 17.0, 40.0],
 [6.0, 2.0, 22.0, 17.0, 76.0],
 [7.0, 2.0, 35.0, 18.0, 6.0],
 [8.0, 2.0, 23.0, 18.0, 94.0],
 [9.0, 1.0, 64.0, 19.0, 3.0],
 [10.0, 2.0, 30.0, 19.0, 72.0],
 [11.0, 1.0, 67.0, 19.0, 14.0],
 [12.0, 2.0, 35.0, 19.0, 99.0],
 [13.0, 2.0, 58.0, 20.0, 15.0],
 [14.0, 2.0, 24.0, 20.0, 77.0],
 [15.0, 1.0, 37.0, 20.0, 13.0],
 [16.0, 1.0, 22.0, 20.0, 79.0],
 [17.0, 2.0, 35.0, 21.0, 35.0],
 [18.0, 1.0, 20.0, 21.0, 66.0],
 [19.0, 1.0, 52.0, 23.0, 29.0],
 [20.0, 2.0, 35.0, 23.0, 98.0],
 [21.0, 1.0, 35.0, 24.0, 35.0],
 [22.0, 1.0, 25.0, 24.0, 73.0],
 [23.0, 2.0, 46.0, 25.0, 5.0],
 [24.0, 1.0, 31.0, 25.0, 73.0],
 [25.0, 2.0, 54.0, 28.0, 14.0],
 [26.0, 1.0, 29.0, 28.0, 82.0],
 [27.0, 2.0, 45.0, 28.0, 32.0],
 [28.0, 1.0, 35.0, 28.0, 61.0],
 [29.0, 2.0, 40.0, 29.0, 31.0],
 [30.0, 2.0, 23.0, 29.0, 87.0],
 [31.0, 1.0, 60.0, 30.0, 4.0],
 [32.0, 2.0, 21.0, 30.

In [22]:
data[2][2]

20.0

In [23]:
data[2][2] + data[3][2]

43.0

## 벡터 뽑기

In [24]:
age = [] 

for i in range(0, n_row-1):  # 한 줄 씩 조회하기.
    val = data[i][2]         # 2가 Age 열의 데이터이기 때문에.
    age.append(val)

In [25]:
age

[19.0,
 21.0,
 20.0,
 23.0,
 31.0,
 22.0,
 35.0,
 23.0,
 64.0,
 30.0,
 67.0,
 35.0,
 58.0,
 24.0,
 37.0,
 22.0,
 35.0,
 20.0,
 52.0,
 35.0,
 35.0,
 25.0,
 46.0,
 31.0,
 54.0,
 29.0,
 45.0,
 35.0,
 40.0,
 23.0,
 60.0,
 21.0,
 53.0,
 18.0,
 49.0,
 21.0,
 42.0,
 30.0,
 36.0,
 20.0,
 65.0,
 24.0,
 48.0,
 31.0,
 49.0,
 24.0,
 50.0,
 27.0,
 29.0,
 31.0,
 49.0,
 33.0,
 31.0,
 59.0,
 50.0,
 47.0,
 51.0,
 69.0,
 27.0,
 53.0,
 70.0,
 19.0,
 67.0,
 54.0,
 63.0,
 18.0,
 43.0,
 68.0,
 19.0,
 32.0,
 70.0,
 47.0,
 60.0,
 60.0,
 59.0,
 26.0,
 45.0,
 40.0,
 23.0,
 49.0,
 57.0,
 38.0,
 67.0,
 46.0,
 21.0,
 48.0,
 55.0,
 22.0,
 34.0,
 50.0,
 68.0,
 18.0,
 48.0,
 40.0,
 32.0,
 24.0,
 47.0,
 27.0,
 48.0,
 20.0,
 23.0,
 49.0,
 67.0,
 26.0,
 49.0,
 21.0,
 66.0,
 54.0,
 68.0,
 66.0,
 65.0,
 19.0,
 38.0,
 19.0,
 18.0,
 19.0,
 63.0,
 49.0,
 51.0,
 50.0,
 27.0,
 38.0,
 40.0,
 39.0,
 23.0,
 31.0,
 43.0,
 40.0,
 59.0,
 38.0,
 47.0,
 39.0,
 25.0,
 31.0,
 20.0,
 29.0,
 44.0,
 32.0,
 19.0,
 35.0,
 57.0,
 32.0,
 28.0,

pandas에는 시리즈라는 자료형으로 나올 것이다. 시리즈와 데이터프레임끼리 연산을 못한다던가 문제가 생긴다. 

그러나 우리가 지금 가져온 것은 리스트라서 연산이 안될 수가 없다. 판다스로 하다가 안될 때 바꿔서 하면 문제 해결! ㅎㅁㅎ

In [26]:
score = []

for i in range(0, n_row-1):
    val = data[i][4]
    score.append(val)

In [27]:
score

[39.0,
 81.0,
 6.0,
 77.0,
 40.0,
 76.0,
 6.0,
 94.0,
 3.0,
 72.0,
 14.0,
 99.0,
 15.0,
 77.0,
 13.0,
 79.0,
 35.0,
 66.0,
 29.0,
 98.0,
 35.0,
 73.0,
 5.0,
 73.0,
 14.0,
 82.0,
 32.0,
 61.0,
 31.0,
 87.0,
 4.0,
 73.0,
 4.0,
 92.0,
 14.0,
 81.0,
 17.0,
 73.0,
 26.0,
 75.0,
 35.0,
 92.0,
 36.0,
 61.0,
 28.0,
 65.0,
 55.0,
 47.0,
 42.0,
 42.0,
 52.0,
 60.0,
 54.0,
 60.0,
 45.0,
 41.0,
 50.0,
 46.0,
 51.0,
 46.0,
 56.0,
 55.0,
 52.0,
 59.0,
 51.0,
 59.0,
 50.0,
 48.0,
 59.0,
 47.0,
 55.0,
 42.0,
 49.0,
 56.0,
 47.0,
 54.0,
 53.0,
 48.0,
 52.0,
 42.0,
 51.0,
 55.0,
 41.0,
 44.0,
 57.0,
 46.0,
 58.0,
 55.0,
 60.0,
 46.0,
 55.0,
 41.0,
 49.0,
 40.0,
 42.0,
 52.0,
 47.0,
 50.0,
 42.0,
 49.0,
 41.0,
 48.0,
 59.0,
 55.0,
 56.0,
 42.0,
 50.0,
 46.0,
 43.0,
 48.0,
 52.0,
 54.0,
 42.0,
 46.0,
 48.0,
 50.0,
 43.0,
 59.0,
 43.0,
 57.0,
 56.0,
 40.0,
 58.0,
 91.0,
 29.0,
 77.0,
 35.0,
 95.0,
 11.0,
 75.0,
 9.0,
 75.0,
 34.0,
 71.0,
 5.0,
 88.0,
 7.0,
 73.0,
 10.0,
 72.0,
 5.0,
 93.0,
 40.0,
 87.0,
 1

pandas로 하지만, 층이 하나씩 올라갈 수록 위험성도 높아지는 것을 인지하는 것이 좋다. 쓸 수 있으면 좋지만, 못쓰는 상황이 올 수도 있다는 것을 생각해야.

In [28]:
res = np.add(age, score)

In [29]:
res

array([ 58., 102.,  26., 100.,  71.,  98.,  41., 117.,  67., 102.,  81.,
       134.,  73., 101.,  50., 101.,  70.,  86.,  81., 133.,  70.,  98.,
        51., 104.,  68., 111.,  77.,  96.,  71., 110.,  64.,  94.,  57.,
       110.,  63., 102.,  59., 103.,  62.,  95., 100., 116.,  84.,  92.,
        77.,  89., 105.,  74.,  71.,  73., 101.,  93.,  85., 119.,  95.,
        88., 101., 115.,  78.,  99., 126.,  74., 119., 113., 114.,  77.,
        93., 116.,  78.,  79., 125.,  89., 109., 116., 106.,  80.,  98.,
        88.,  75.,  91., 108.,  93., 108.,  90.,  78.,  94., 113.,  77.,
        94.,  96., 123.,  59.,  97.,  80.,  74.,  76.,  94.,  77.,  90.,
        69.,  64.,  97., 126.,  81., 105.,  63., 116., 100., 111., 114.,
       117.,  73.,  80.,  65.,  66.,  69., 106., 108.,  94., 107.,  83.,
        78.,  98., 130.,  52., 108.,  78., 135.,  70., 113.,  56., 114.,
        59., 102.,  25., 117.,  51., 105.,  29., 107.,  62., 125.,  68.,
       119.,  37., 125.,  84., 106.,  56., 124.,  6

우리가 리스트와 리스트 즉 벡터와 벡터를 넣었더니 결과가 바로 나온다는 것이에요.

In [30]:
type(res)

numpy.ndarray

두 개를 더했더니 numpy.ndarray 라는 자료형이라는 것이다. 

darray라는 자료형은 numpy의 자료형이다. 이대로 작업하면 또 이것떄문에 문제가 생길 수 있다.

벡터는 무조건 리스트! 그 리스트로 다 할 수 있어야 한다.

## darray 형태의 벡터를 list로 변환하기

In [31]:
res.tolist()

[58.0,
 102.0,
 26.0,
 100.0,
 71.0,
 98.0,
 41.0,
 117.0,
 67.0,
 102.0,
 81.0,
 134.0,
 73.0,
 101.0,
 50.0,
 101.0,
 70.0,
 86.0,
 81.0,
 133.0,
 70.0,
 98.0,
 51.0,
 104.0,
 68.0,
 111.0,
 77.0,
 96.0,
 71.0,
 110.0,
 64.0,
 94.0,
 57.0,
 110.0,
 63.0,
 102.0,
 59.0,
 103.0,
 62.0,
 95.0,
 100.0,
 116.0,
 84.0,
 92.0,
 77.0,
 89.0,
 105.0,
 74.0,
 71.0,
 73.0,
 101.0,
 93.0,
 85.0,
 119.0,
 95.0,
 88.0,
 101.0,
 115.0,
 78.0,
 99.0,
 126.0,
 74.0,
 119.0,
 113.0,
 114.0,
 77.0,
 93.0,
 116.0,
 78.0,
 79.0,
 125.0,
 89.0,
 109.0,
 116.0,
 106.0,
 80.0,
 98.0,
 88.0,
 75.0,
 91.0,
 108.0,
 93.0,
 108.0,
 90.0,
 78.0,
 94.0,
 113.0,
 77.0,
 94.0,
 96.0,
 123.0,
 59.0,
 97.0,
 80.0,
 74.0,
 76.0,
 94.0,
 77.0,
 90.0,
 69.0,
 64.0,
 97.0,
 126.0,
 81.0,
 105.0,
 63.0,
 116.0,
 100.0,
 111.0,
 114.0,
 117.0,
 73.0,
 80.0,
 65.0,
 66.0,
 69.0,
 106.0,
 108.0,
 94.0,
 107.0,
 83.0,
 78.0,
 98.0,
 130.0,
 52.0,
 108.0,
 78.0,
 135.0,
 70.0,
 113.0,
 56.0,
 114.0,
 59.0,
 102.0,
 25.0,
 117.

array형식이 좋다면

In [32]:
A = np.array(age)
B = np.array(score)
C = A + B
C

array([ 58., 102.,  26., 100.,  71.,  98.,  41., 117.,  67., 102.,  81.,
       134.,  73., 101.,  50., 101.,  70.,  86.,  81., 133.,  70.,  98.,
        51., 104.,  68., 111.,  77.,  96.,  71., 110.,  64.,  94.,  57.,
       110.,  63., 102.,  59., 103.,  62.,  95., 100., 116.,  84.,  92.,
        77.,  89., 105.,  74.,  71.,  73., 101.,  93.,  85., 119.,  95.,
        88., 101., 115.,  78.,  99., 126.,  74., 119., 113., 114.,  77.,
        93., 116.,  78.,  79., 125.,  89., 109., 116., 106.,  80.,  98.,
        88.,  75.,  91., 108.,  93., 108.,  90.,  78.,  94., 113.,  77.,
        94.,  96., 123.,  59.,  97.,  80.,  74.,  76.,  94.,  77.,  90.,
        69.,  64.,  97., 126.,  81., 105.,  63., 116., 100., 111., 114.,
       117.,  73.,  80.,  65.,  66.,  69., 106., 108.,  94., 107.,  83.,
        78.,  98., 130.,  52., 108.,  78., 135.,  70., 113.,  56., 114.,
        59., 102.,  25., 117.,  51., 105.,  29., 107.,  62., 125.,  68.,
       119.,  37., 125.,  84., 106.,  56., 124.,  6

# 전치행렬 구하기

In [33]:
data_t = np.transpose(data)

In [34]:
data_t

array([[  1.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,   9.,  10.,  11.,
         12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,  21.,  22.,
         23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,  33.,
         34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  43.,  44.,
         45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.,  55.,
         56.,  57.,  58.,  59.,  60.,  61.,  62.,  63.,  64.,  65.,  66.,
         67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,  75.,  76.,  77.,
         78.,  79.,  80.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,  88.,
         89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  97.,  98.,  99.,
        100., 101., 102., 103., 104., 105., 106., 107., 108., 109., 110.,
        111., 112., 113., 114., 115., 116., 117., 118., 119., 120., 121.,
        122., 123., 124., 125., 126., 127., 128., 129., 130., 131., 132.,
        133., 134., 135., 136., 137., 138., 139., 140., 141., 142., 143.,
        144., 145., 146., 147., 148., 

## XtX 구하기

In [35]:
XtX = data_t.dot(data)

In [36]:
XtX

array([[2.686700e+06, 3.102700e+04, 7.765790e+05, 1.512980e+06,
        1.013135e+06],
       [3.102700e+04, 5.360000e+02, 1.203700e+04, 1.874800e+04,
        1.581100e+04],
       [7.765790e+05, 1.203700e+04, 3.406960e+05, 4.696460e+05,
        3.665640e+05],
       [1.512980e+06, 1.874800e+04, 4.696460e+05, 8.707800e+05,
        6.093590e+05],
       [1.013135e+06, 1.581100e+04, 3.665640e+05, 6.093590e+05,
        6.367120e+05]])

In [37]:
XtX.tolist()

[[2686700.0, 31027.0, 776579.0, 1512980.0, 1013135.0],
 [31027.0, 536.0, 12037.0, 18748.0, 15811.0],
 [776579.0, 12037.0, 340696.0, 469646.0, 366564.0],
 [1512980.0, 18748.0, 469646.0, 870780.0, 609359.0],
 [1013135.0, 15811.0, 366564.0, 609359.0, 636712.0]]

XtX 는 대칭행렬이다.